homeassistant/components/hassio/__init__.py

```diff
@@ -455,4 +475,138 @@ async def async_handle_core_service(call):
+
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
+
     return True
+
+
+async def async_unload_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry
+) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+
+    # Pop add-on data
+    hass.data.pop(ADDONS_COORDINATOR, None)
+
+    return unload_ok
+
+
+@callback
+def async_register_addons_in_dev_reg(
+    entry_id: str, dev_reg: DeviceRegistry, addons: List[Dict[str, Any]]
+) -> None:
+    """Register addons in the device registry."""
+    for addon in addons:
+        dev_reg.async_get_or_create(
+            config_entry_id=entry_id,
+            identifiers={(DOMAIN, addon[ATTR_SLUG])},
+            manufacturer=addon.get(ATTR_REPOSITORY) or addon.get(ATTR_URL) or "unknown",
+            model="Home Assistant Add-on",
+            sw_version=addon[ATTR_VERSION],
+            name=addon[ATTR_NAME],
+            entry_type=ATTR_SERVICE,
+        )
+
+
+@callback
+def async_register_os_in_dev_reg(
+    entry_id: str, dev_reg: DeviceRegistry, os_dict: Dict[str, Any]
+) -> None:
+    """Register OS in the device registry."""
+    dev_reg.async_get_or_create(
+        config_entry_id=entry_id,
+        identifiers={(DOMAIN, "OS")},
+        manufacturer="Home Assistant",
+        model="Home Assistant Operating System",
+        sw_version=os_dict[ATTR_VERSION],
+        name="Home Assistant Operating System",
+        entry_type=ATTR_SERVICE,
+    )
+
+
+@callback
+def async_remove_addons_from_dev_reg(
+    dev_reg: DeviceRegistry, addons: List[Dict[str, Any]]
+) -> None:
+    """Remove addons from the device registry."""
+    for addon_slug in addons:
+        if dev := dev_reg.async_get_device({(DOMAIN, addon_slug)}):
+            dev_reg.async_remove_device(dev.id)
+
+
+class HassioDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to retrieve Hass.io status."""
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, dev_reg: DeviceRegistry
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self._async_update_data,
+        )
+        self.data = {}
+        self.entry_id = config_entry.entry_id
+        self.dev_reg = dev_reg
+
+    async def _async_update_data(self) -> Dict[str, Any]:
+        """Update data via library."""
+        new_data = {}
+        addon_data = get_supervisor_info(self.hass)
+
+        new_data["addons"] = {
+            addon[ATTR_SLUG]: addon for addon in addon_data.get("addons", [])
+        }
+        new_data["os"] = get_os_info(self.hass)
```

Comment:
This should be limited to Home Assistant Operating System instances

Reply:
how can I determine that the instance is an HA OS one?

Reply:
<https://github.com/raman325/home-assistant/blob/dev/homeassistant/components/hassio/system_health.py#L54> This is how we do it for system_health